src/encoder.rs

```diff
@@ -504,17 +504,29 @@ impl<T: Pixel> FrameInvariants<T> {
     let w_in_b = 2 * config.width.align_power_of_two_and_shift(3); // MiCols, ((width+7)/8)<<3 >> MI_SIZE_LOG2
     let h_in_b = 2 * config.height.align_power_of_two_and_shift(3); // MiRows, ((height+7)/8)<<3 >> MI_SIZE_LOG2
 
+    let frame_rate = config.time_base.den / config.time_base.num;
+    let min_tile_cols = (config.width - 1) / 4096;
+    // The minimum number of tiles is determined based on the following requirements:
+    // - A tile cannot be wider than 4096 pixels
+    // - A tile cannot be larger than 4096x2304 pixels
+    // - The tile size * the frame rate * a temporal ratio cannot exceed a given fixed rate
+    //   corresponding to 4K60 resolution with a 1.1 ratio.
+    let min_tiles = 1 + min_tile_cols.max(
+      (config.width * config.height) / (4096 * 2304)).max(
+      (((config.width * config.height) as u64 * frame_rate) as f64 * 1.1) as usize / (4096 * 2176 * 60) as usize
+    );
+
     let mut tiling = TilingInfo::new(
       sequence.sb_size_log2(),
       config.width,
       config.height,
-      config.tile_cols_log2,
+      config.tile_cols_log2.min(min_tile_cols),
       config.tile_rows_log2
     );
 
-    if config.tiles > 0 {
+    if config.tiles > 0 || tiling.rows * tiling.cols < min_tiles {
       let mut tile_rows_log2 = 0;
```

Comment:
It might make sense to adjust the behavior slightly here. This was written normally so that if you specify --tiles, tile_rows_log2 is ignored. However, if we failed to meet max tile size, we maybe don't want to totally ignore those, but treat them as a minimum. So maybe set tile_rows_log2 to config.tile_rows_log2 here, but only if (config.tiles == 0 && tiling.rows * tiling.cols < min_tiles)

Reply:
(Another option is just to fail here if the user uses tile_rows_log2/cols_log2 explicitly rather than --tiles)

Reply:
Agreed; should be fixed now.